src/pil/backend/powdr_pil.rs

```diff
@@ -0,0 +1,488 @@
+) {
+    if let Some(lookups) = pil_ir.lookups.get(step_type) {
+        for lookup in lookups {
+            let mut lookup_source: Vec<String> = Vec::new();
+            let mut lookup_destination: Vec<String> = Vec::new();
+            for (src, dest) in lookup {
+                lookup_source.push(convert_to_pil_expr_string(src.clone(), annotations_map));
+                lookup_destination.push(convert_to_pil_expr_string(dest.clone(), annotations_map));
+            }
+            // PIL lookups have the format of `step_type { lookup_sources } in {
+            // lookup_destinations }`.
+            writeln!(
+                pil,
+                "{} {{{}}} in {{{}}} ",
+                step_type_name,
+                lookup_source.join(", "),
+                lookup_destination.join(", ")
+            )
+            .unwrap();
+        }
+    }
+}
+
+// Convert expression to PIL string recursively. Coding this up separately because PIL has different
+// syntax for queries.
+fn convert_to_pil_expr_string<F: Debug + Clone>(
+    expr: Expr<F, Queriable<F>>,
+    annotations_map: &HashMap<UUID, String>,
+) -> String {
+    match expr {
+        Expr::Const(constant) => format!("{:?}", constant),
+        Expr::Sum(sum) => {
+            let mut expr_string = String::new();
+            for (index, expr) in sum.iter().enumerate() {
+                expr_string += convert_to_pil_expr_string(expr.clone(), annotations_map).as_str();
+                if index != sum.len() - 1 {
+                    expr_string += " + ";
+                }
+            }
+            format!("({})", expr_string)
+        }
+        Expr::Mul(mul) => {
+            let mut expr_string = String::new();
+            for (index, expr) in mul.iter().enumerate() {
+                expr_string += convert_to_pil_expr_string(expr.clone(), annotations_map).as_str();
+                if index != mul.len() - 1 {
+                    expr_string += " * ";
+                }
+            }
+            format!("({})", expr_string)
+        }
+        Expr::Neg(neg) => format!("(-{})", convert_to_pil_expr_string(*neg, annotations_map)),
+        Expr::Pow(pow, power) => {
+            format!(
+                "({})^{}",
+                convert_to_pil_expr_string(*pow, annotations_map),
+                power
+            )
+        }
+        Expr::Query(queriable) => convert_to_pil_queriable_string(queriable, annotations_map),
+        Expr::Halo2Expr(_) => {
+            panic!("Halo2 native expression not supported by PIL backend.")
+        }
+    }
+}
+
+// Convert queriable to PIL string recursively. Major differences are: 1. PIL rotations are in the
+// format of an apostrophe after the signal; 2. PIL only supports the next rotation, so there's no
+// previous or arbitrary rotation.
+fn convert_to_pil_queriable_string<F>(
+    query: Queriable<F>,
+    annotations_map: &HashMap<UUID, String>,
+) -> String {
+    match query {
```

Comment:
We should have a column type already in PIL IR, this translation should be done in the compiler

Reply:
I see your point. I can create a column and query system in PIL IR, though it'll be almost identical to our Queriable, maybe except for the rotation part.